# Student Performance Indicator

Life cycle of Machine Learning Project

- understanding the problem
- data collection
- data checks to perform
- exploratory data analysis
- model training
- choose the best model

## 1. Problem Statement
- This project understands how the student's performance (test scores) is affected by other variables such as Gender, Ethnicity, Parental level of education, Lunch and Test preparation course.

## 2. Data Collectionn
- Data Scouce: click [here](https://www.kaggle.com/datasets/spscientist/students-performance-in-exams?datasetId=74977) to navigate to data source in kaggle.
- The data has 8 columns and 1000 rows.

### 2.1 Import data and required packages
**Importing Pandas, NumPy, Matplotlib, Seaborn and Warnings Library**

In [34]:
import numpy as np
import pandas as pd
import seaborn as sns
import json
import matplotlib.pyplot as plt
from string_cleaner import clean_string
import warnings

warnings.filterwarnings("ignore")

**Import the CSV data as pandas dataframe**

In [35]:
students = pd.read_csv("./data/students.csv")

In [36]:
students

,gender,race/ethnicity,parental level of education,lunch,test preparation course,math score,reading score,writing score
0,female,group B,bachelor's degree,standard,none,72,72,74
1,female,group C,some college,standard,completed,69,90,88
2,female,group B,master's degree,standard,none,90,95,93
3,male,group A,associate's degree,free/reduced,none,47,57,44
4,male,group C,some college,standard,none,76,78,75
...,...,...,...,...,...,...,...,...
995,female,group E,master's degree,standard,completed,88,99,95
996,male,group C,high school,free/reduced,none,62,55,55
997,female,group C,high school,free/reduced,completed,59,71,65
998,female,group D,some college,standard,completed,68,78,77


**Clean the column names**

In [37]:
students = students.rename(
    columns=clean_string
)

In [38]:
students

,gender,race_ethnicity,parental_level_of_education,lunch,test_preparation_course,math_score,reading_score,writing_score
0,female,group B,bachelor's degree,standard,none,72,72,74
1,female,group C,some college,standard,completed,69,90,88
2,female,group B,master's degree,standard,none,90,95,93
3,male,group A,associate's degree,free/reduced,none,47,57,44
4,male,group C,some college,standard,none,76,78,75
...,...,...,...,...,...,...,...,...
995,female,group E,master's degree,standard,completed,88,99,95
996,male,group C,high school,free/reduced,none,62,55,55
997,female,group C,high school,free/reduced,completed,59,71,65
998,female,group D,some college,standard,completed,68,78,77


In [39]:
students.shape

(1000, 8)

### 2.2 Dataset Information
| Column Name                 | Description                                   | Possible Values                                                                   |
|:----------------------------|:----------------------------------------------|:----------------------------------------------------------------------------------|
| gender                      | Sex of the student                            | Male, Female                                                                      |
| race/ethnicity              | Ethnicity group of the student                | Group A, Group B, Group C, Group D, Group E                                       |
| parental level of education | Parents' highest level of education           | Bachelor's degree, Some college, Master's degree, Associate's degree, High school |
| lunch                       | Whether the student had lunch before the test | Standard, Free/Reduced                                                            |
| test preparation course     | Completion status of test preparation course  | Completed, Not completed                                                          |
| math score                  | Score obtained in math                        | Numeric (0–100)                                                                   |
| reading score               | Score obtained in reading                     | Numeric (0–100)                                                                   |
| writing score               | Score obtained in writing                     | Numeric (0–100)                                                                   |

## 3. Data check to Perform
- Check missing values
- Check for duplicates
- Check data types
- Check unique values in each column
- Check for statistics of data set
- Check different categories in various categorical columns


### 3.1 Check missing values

In [40]:
students.isnull().sum()

gender                         0
race_ethnicity                 0
parental_level_of_education    0
lunch                          0
test_preparation_course        0
math_score                     0
reading_score                  0
writing_score                  0
dtype: int64

**Fortunately there are no null values in the data set**

### 3.2 Check for duplicates

In [41]:
students.duplicated().sum()

np.int64(0)

**There are no duplcates in the dataset**

### 3.3 Check data types

In [42]:
students.info()

<class 'pandas.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype
---  ------                       --------------  -----
 0   gender                       1000 non-null   str  
 1   race_ethnicity               1000 non-null   str  
 2   parental_level_of_education  1000 non-null   str  
 3   lunch                        1000 non-null   str  
 4   test_preparation_course      1000 non-null   str  
 5   math_score                   1000 non-null   int64
 6   reading_score                1000 non-null   int64
 7   writing_score                1000 non-null   int64
dtypes: int64(3), str(5)
memory usage: 62.6 KB


### 3.4 Checking the number of values in each columns

In [43]:
students.nunique()

gender                          2
race_ethnicity                  5
parental_level_of_education     6
lunch                           2
test_preparation_course         2
math_score                     81
reading_score                  72
writing_score                  77
dtype: int64

**The above table shows how each column have how many duplicates**

### 3.5 Check statistics of data

In [44]:
students.describe(
    include = "all"
)

,gender,race_ethnicity,parental_level_of_education,lunch,test_preparation_course,math_score,reading_score,writing_score
count,1000,1000,1000,1000,1000,1000.00000,1000.000000,1000.000000
unique,2,5,6,2,2,NaN,NaN,NaN
top,female,group C,some college,standard,none,NaN,NaN,NaN
freq,518,319,226,645,642,NaN,NaN,NaN
mean,NaN,NaN,NaN,NaN,NaN,66.08900,69.169000,68.054000
std,NaN,NaN,NaN,NaN,NaN,15.16308,14.600192,15.195657
min,NaN,NaN,NaN,NaN,NaN,0.00000,17.000000,10.000000
25%,NaN,NaN,NaN,NaN,NaN,57.00000,59.000000,57.750000
50%,NaN,NaN,NaN,NaN,NaN,66.00000,70.000000,69.000000
75%,NaN,NaN,NaN,NaN,NaN,77.00000,79.000000,79.000000


#### Insights:
- There are three numerical columns: `math_score`, `reading_score`, `writing_score`.
- The means of all these three columns are very much close to each other: between `66.08` and `69.19`.
- The standard deviations of all these numerical columns are very much close too: between `14.60` and `15.19`.
- Math Score has a minimum of `0.00`, Reading Score has a minimum of `17.00` while Writing Score has a minimum of `10.00`.

### 3.6 Exploring Data

In [45]:
students.head()

,gender,race_ethnicity,parental_level_of_education,lunch,test_preparation_course,math_score,reading_score,writing_score
0,female,group B,bachelor's degree,standard,none,72,72,74
1,female,group C,some college,standard,completed,69,90,88
2,female,group B,master's degree,standard,none,90,95,93
3,male,group A,associate's degree,free/reduced,none,47,57,44
4,male,group C,some college,standard,none,76,78,75


In [49]:
categories = {column_name: list(students[column_name].unique()) for column_name in students.select_dtypes(exclude=["number"]).columns}

In [50]:
print("Shown below are the different categories for each variable in students dataframe:")
print(json.dumps(categories, indent=4))

Shown below are the different categories for each variable in students dataframe:
{
    "gender": [
        "female",
        "male"
    ],
    "race_ethnicity": [
        "group B",
        "group C",
        "group A",
        "group D",
        "group E"
    ],
    "parental_level_of_education": [
        "bachelor's degree",
        "some college",
        "master's degree",
        "associate's degree",
        "high school",
        "some high school"
    ],
    "lunch": [
        "standard",
        "free/reduced"
    ],
    "test_preparation_course": [
        "none",
        "completed"
    ]
}
